### PR TITLE
Refactor Product Recommendation sample notebook 

### PR DESCRIPTION
**Pull request by jainr**
## Description
Our samples had several usability issues, specially with lot of hardcoded values and confusing environment variable and config file setting, this PR addresses lot of those issues. 

Once we get feedback on this PR, I will update other samples.

## How was this PR tested?

Ran the notebook successfully on AML

image

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to clarify your proposed changes.

### CODE REVIEW

File: docs/samples/product_recommendation_demo.ipynb [hunk 1]
```diff
@@ -200,68 +193,38 @@
...
         "  azure_synapse:\n",
-        "    dev_url: 'https://feathrazuretest3synapse.dev.azuresynapse.net'\n",
-        "    pool_name: 'spark3'\n",
-        "    workspace_dir: 'abfss://feathrazuretest3fs@feathrazuretest3storage.dfs.core.windows.net/feathr_getting_started'\n",
+        "    dev_url: 'https://{SYNAPSE_WORKSPACE_NAME}.dev.azuresynapse.net'\n",
+        "    pool_name: 'spark31'\n",
+        "    workspace_dir: 'abfss://{ADLS_FS_NAME}@{ADLS_ACCOUNT}.dfs.core.windows.net/feathr_project'\n",
         "    executor_size: 'Small'\n",
         "    executor_num: 1\n",
-        "  databricks:\n",
```
Comment (Yuqing-cat) [inline docs/samples/product_recommendation_demo.ipynb:237]: Not sure if I understand it correctly. Databricks settings are removed, so this notebook will be a synapse e2e sample, right? If so, do we need to move this sample to a `synapse` subfolder like Databricks sample do?
https://github.com/feathr-ai/feathr/tree/main/docs/samples/databricks

File: docs/samples/product_recommendation_demo.ipynb [hunk 2]
```diff
@@ -4,38 +4,40 @@
       "cell_type": "markdown",
       "metadata": {},
       "source": [
-        "# Demo Notebook: Feathr Feature Store on Azure\n",
+        "# Product Recommendation with Feathr\n",
```
Comment (Yuqing-cat) [inline docs/samples/product_recommendation_demo.ipynb:7]: `on Azure` is removed from title, but it's still useful to readers of an open source project to know this sample requires Azure subscription.

File: docs/samples/product_recommendation_demo.ipynb [hunk 3]
```diff
@@ -4,38 +4,40 @@
...
         "\n",
         "This notebook demonstrates how Feathr Feature Store can simplify and empower your model training and inference. You will learn:\n",
         "\n",
         "1. Define sharable features using Feathr API\n",
         "2. Create a training dataset via point-in-time feature join with Feathr API\n",
-        "3. Materialize features to online store and then retrieve them with Feathr API"
+        "3. Materialize features to online store and then retrieve them with Feathr API\n",
```
Comment (Yuqing-cat) [inline docs/samples/product_recommendation_demo.ipynb:13]: `4. register feature with register API` could be added according to the notebook content. 

And I would like to propose moving `register feature step` to step 2. Users are usually confused and think the feature registry are related to spark environment.

File: docs/samples/product_recommendation_demo.ipynb (hunk 1 above)
Comment (jainr) [reply docs/samples/product_recommendation_demo.ipynb:237]: Good idea, moved it. Once this PR merges, I will make PRs for other notebooks too. May be we need to add similar notebooks for databricks too.

File: docs/samples/product_recommendation_demo.ipynb (hunk 2 above)
Comment (jainr) [reply docs/samples/product_recommendation_demo.ipynb:7]: Added Azure back

File: docs/samples/product_recommendation_demo.ipynb (hunk 3 above)
Comment (jainr) [reply docs/samples/product_recommendation_demo.ipynb:13]: Great callout, addressed

File: docs/samples/azure_synapse/product_recommendation_demo.ipynb [hunk 4]
```diff
@@ -270,10 +234,10 @@
       "source": [
         "# Define sharable features using Feathr API\n",
         "\n",
-        "In this tutorial, we use Feathr Feature Store to help create a model that predicts users product rating. To make it simple, let's just predict users' rating for ONE product for an e-commerce website. (We have an [advanced demo](./product_recommendation_demo_advanced.ipynb) that predicts ratings for arbitrary products.)\n",
+        "In this tutorial, we use Feathr Feature Store and create a model that predicts users' product rating. To make it simple, let's just predict users' rating for ONE product for an e-commerce website. (We have an [advanced demo](./product_recommendation_demo_advanced.ipynb) that predicts ratings for arbitrary products.)\n",
```
Comment (blrchen) [inline docs/samples/azure_synapse/product_recommendation_demo.ipynb:237]: This advance nb link will be broken after this PR merged.

**Comment by blrchen** [comment]
>         "if feathr_client.spark_runtime == 'databricks':\n",

Since this notebook is for synpase only, it's better to remove databricks related content to avoid confusion

---
Refers to: docs/samples/azure_synapse/product_recommendation_demo.ipynb:541 in 6419344. [](commit_id = 641934476ca10b7d285835fbc2322ae5980853df, deletion_comment = False)

File: docs/samples/azure_synapse/product_recommendation_demo.ipynb (hunk 4 above)
Comment (jainr) [reply docs/samples/azure_synapse/product_recommendation_demo.ipynb]: Fixed, thanks for pointing it out.

File: docs/samples/azure_synapse/product_recommendation_demo.ipynb [hunk 5]
```diff
@@ -694,16 +653,16 @@
         "\n",
         "In the previous section, we demonstrated how Feathr can compute feature value to generate training dataset from feature definition on-they-fly.\n",
         "\n",
-        "Now let's talk about how we can use the trained models. We can use the trained models for offline inference as well as online inference. In both cases, we need features to be feed into the models. For offline inference, you can compute and get the features on-demand; or you can store the computed features to some offline database for later offline inference.\n",
+        "Now let's talk about how we can use the trained models. We can use the trained models for both online and offline inference. In both cases, we need features to be fed into the models. For offline inference, you can compute and get the features on-demand; or you can store the computed features to some offline database for later offline inference.\n",
         "\n",
         "For online inference, we can use Feathr to compute and store the features in the online database. Then use it for online inference when the request comes.\n",
         "\n",
         "![img](../images/online_inference.jpg)\n",
```
[the remark this answers is not in the document]
Comment (jainr) [reply docs/samples/azure_synapse/product_recommendation_demo.ipynb]: Fixed, thanks for pointing it out.

**Comment by jainr** [comment]
> > ```
> >     "if feathr_client.spark_runtime == 'databricks':\n",
> > ```
> 
> Since this notebook is for synpase only, it's better to remove databricks related content to avoid confusion
> 
> Refers to: docs/samples/azure_synapse/product_recommendation_demo.ipynb:541 in [6419344](https://github.com/feathr-ai/feathr/commit/641934476ca10b7d285835fbc2322ae5980853df). [](commit_id = [6419344](https://github.com/feathr-ai/feathr/commit/641934476ca10b7d285835fbc2322ae5980853df), deletion_comment = False)

Fixed.

**Review by Yuqing-cat — APPROVED** [review]

**Review by Yuqing-cat — APPROVED** [review]

**Review by blrchen — APPROVED** [review]

**Pull request merged**